### ISSUE TITLE
Reduce developer time spent debugging "why mailchimp sync didn't complete".

### ISSUE THREAD

**Issue by elyalvarado** [elyalvarado/advent-of-code-2018#10]
Ever since we introduced the Smile Mailchimp app, we’ve had merchant reports of the Smile → Mailchimp sync “not completing” (the trackable tasks completes, but does not syncing all of the merchant’s customers to Mailchimp). Debugging this with the information we currently have available to us is both time intensive and requires a detailed understanding of both how Smile’s sync works and how Mailchimp’s batch processing works, requires making manual API calls to Mailchimp, and even then we’re usually only able to suggest what*might*have happened, not definitively say what went wrong. This is frustrating to all involved (developers, support, and the merchant).

Some examples of why this is complicated:

Smile’s Mailchimp sync makes use of Mailchimp’s batch API. This API only returns a count of failed records, any details about specifically*which*records failed and why are only available via a gzipped file (the URL of which is included in this API response). Smile does not currently download or store these files or their contents, and batches are only kept by Mailchimp for 7 days. This means that by the time the merchant reports an issue and a developer has time to investigate the detailed context of which customers failed to import*may be*lost forever.

Mailchimp has a known issue where customer records may exist*on mailchimp*in an invalid state (usually because the merchant has configured a merge field as “required” but not all customers have a value for the field). In this case, Smile is*unable*to sync data to these customers, as Mailchimp will throw an “invalid record” error when we try to update these customers, even though the field that is making the record invalid has nothing to do with Smile or the API request Smile is trying to make.

tl:dr; The Smile Mailchimp app is frequently a pain-point for large merchants, Smile Support, and Smile developers, and it’s anticipated that making the required changes to better expose why it behaves the way it does in order to make it “simple” and “self-serve” will require substantial effort.

This task is to add whatever instrumentation we possibly can to help diagnose or at least shine a light on some of these issues we face with incomplete mailchimp syncs.
